### PR TITLE
first version of solution

### PR DESCRIPTION
**Pull request by Sashk52**

### CODE REVIEW

File: src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    }
+
+    @Override
+    public List<CinemaHall> getAll() {
+        try {
+            Session session = HibernateUtil.getSessionFactory().openSession();
+            Query<CinemaHall> getAllCinemaHallsQuery = session.createQuery("from CinemaHall",
+                    CinemaHall.class);
+            return getAllCinemaHallsQuery.getResultList();
```
Comment (AndreiYazadzhi) [inline src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java:42]: You can return your list without local variable,but it`s up to you)

File: src/main/java/com/dev/movietheatre/model/MovieSession.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+
+@Entity
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne (fetch = FetchType.LAZY)
+    @JoinColumn(name = "Movie_id")
```
Comment (AndreiYazadzhi) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java]: Sql naming must be in lower case

File: src/main/java/com/dev/movietheatre/model/MovieSession.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    private Movie movie;
+    @ManyToOne (fetch = FetchType.LAZY)
+    @JoinColumn(name = "CinemaHall_id")
```
Comment (AndreiYazadzhi) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java]: Same

File: src/main/java/com/dev/movietheatre/service/impl/CinemaHallServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import com.dev.movietheatre.dao.CinemaHallDao;
+import com.dev.movietheatre.lib.Inject;
+import com.dev.movietheatre.lib.Service;
+import com.dev.movietheatre.model.CinemaHall;
+import com.dev.movietheatre.service.CinemaHallService;
+import java.util.List;
+
+@Service
+public class CinemaHallServiceImpl implements CinemaHallService {
+
```
Comment (AndreiYazadzhi) [inline src/main/java/com/dev/movietheatre/service/impl/CinemaHallServiceImpl.java]: Redundant empty line

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import com.dev.movietheatre.lib.Inject;
+import com.dev.movietheatre.lib.Service;
+import com.dev.movietheatre.model.MovieSession;
+import com.dev.movietheatre.service.MovieSessionService;
+import java.time.LocalDate;
+import java.util.List;
+
+@Service
+public class MovieSessionServiceImpl implements MovieSessionService {
+
```
Comment (AndreiYazadzhi) [inline src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: Same

**Review by AndreiYazadzhi — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java (hunk 5 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/service/impl/CinemaHallServiceImpl.java (hunk 4 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/CinemaHallServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 3 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: done

File: src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java (hunk 1 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java:42]: thanks for advice

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 2 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: done

**Review by AndreiYazadzhi — APPROVED** [review]

File: src/main/java/com/dev/movietheatre/Main.java [hunk 6]
```diff
@@ -1,18 +1,43 @@
...
+import com.dev.movietheatre.service.CinemaHallService;
 import com.dev.movietheatre.service.MovieService;
+import com.dev.movietheatre.service.MovieSessionService;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.List;
 
 public class Main {
     private static Injector injector = Injector.getInstance("com.dev.movietheatre");
+    private static final LocalDate LOCAL_DATE = LocalDate.of(2021,02,01);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]: I think we don't need to make it a constant

File: src/main/java/com/dev/movietheatre/Main.java [hunk 7]
```diff
@@ -1,18 +1,43 @@
...
+        CinemaHall cinemaHall = new CinemaHall();
+        cinemaHall.setDescription("Green");
+        cinemaHallService.add(cinemaHall);
+        cinemaHallService.getAll().forEach(System.out::println);
+
+        MovieSessionService movieSessionService = (MovieSessionService) injector
+                .getInstance(MovieSessionService.class);
+        MovieSession movieSession = new MovieSession();
+        movieSession.setMovie(movie);
+        LocalDateTime localDateTime = LocalDateTime.of(2021,02,01,15,00);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]:
```suggestion
        LocalDateTime localDateTime = LocalDateTime.of(2021, 02, 01, 15, 00);
```

File: src/main/java/com/dev/movietheatre/Main.java [hunk 8]
```diff
@@ -1,18 +1,43 @@
...
+        movieSession.setShowTime(localDateTime);
+        movieSessionService.add(movieSession);
+        List<MovieSession> availableSessions = movieSessionService.findAvailableSessions(1L,
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]: let's use `movie.getId()` instead

File: src/main/java/com/dev/movietheatre/dao/CinemaHallDao.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.dev.movietheatre.dao;
+
+import com.dev.movietheatre.model.CinemaHall;
+import java.util.List;
+
+public interface CinemaHallDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/dao/CinemaHallDao.java]: redundant empty line

File: src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public List<CinemaHall> getAll() {
+        try {
+            Session session = HibernateUtil.getSessionFactory().openSession();
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java]: is the session being closed?

File: src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert cinemaHall entity" + movieSession, e);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java]: why cinemaHall?

File: src/main/java/com/dev/movietheatre/model/CinemaHall.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
+package com.dev.movietheatre.model;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Entity
+public class CinemaHall {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/CinemaHall.java:11]: add `@Table` annotation and configure `proper_table_name`

File: src/main/java/com/dev/movietheatre/model/MovieSession.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+
+@Entity
+public class MovieSession {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java:15]: same

File: src/main/java/com/dev/movietheatre/model/MovieSession.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne (fetch = FetchType.LAZY)
+    @JoinColumn(name = "movie_id")
+    private Movie movie;
+    @ManyToOne (fetch = FetchType.LAZY)
+    @JoinColumn(name = "cinemaHall_id")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/model/MovieSession.java]:
```suggestion
    @JoinColumn(name = "cinema_hall_id")
```

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+
+@Service
+public class MovieSessionServiceImpl implements MovieSessionService {
+    @Inject
+    private MovieSessionDao movieSessionDao;
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: redundant empty line

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        return movieSessionDao.findAvailableSessions(movieId, date);
+    }
+
+    @Override
+    public MovieSession add(MovieSession session) {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/movietheatre/Main.java (hunk 6 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: done

File: src/main/java/com/dev/movietheatre/Main.java (hunk 7 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: done

File: src/main/java/com/dev/movietheatre/Main.java (hunk 8 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: done

File: src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java (hunk 10 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/dao/impl/CinemaHallDaoImpl.java]: no. I replaced by try with resourses.

File: src/main/java/com/dev/movietheatre/dao/CinemaHallDao.java (hunk 9 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/dao/CinemaHallDao.java]: done

File: src/main/java/com/dev/movietheatre/model/CinemaHall.java (hunk 12 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/CinemaHall.java:11]: done

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 13 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java:15]: done

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java (hunk 15 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java (hunk 16 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/service/impl/MovieSessionServiceImpl.java]: done

File: src/main/java/com/dev/movietheatre/model/MovieSession.java (hunk 14 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/model/MovieSession.java]: done

File: src/main/java/com/dev/movietheatre/Main.java [hunk 17]
```diff
@@ -1,18 +1,44 @@
...
+        cinemaHallService.getAll().forEach(System.out::println);
+
+        MovieSessionService movieSessionService = (MovieSessionService) injector
+                .getInstance(MovieSessionService.class);
+        MovieSession movieSession = new MovieSession();
+        movieSession.setMovie(movie);
+        LocalDateTime localDateTime = LocalDateTime.of(2021, 02, 01, 15, 00);
+        movieSession.setShowTime(localDateTime);
+        movieSessionService.add(movieSession);
+        LocalDate movieDate = LocalDate.of(2021,02,01);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/Main.java]:
```suggestion
        LocalDate movieDate = LocalDate.of(2021, 02, 01);
```

File: src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java (hunk 11 above)
Comment (KseniiaMakarova) [inline src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java]: let's make this more informative

**Review by KseniiaMakarova — APPROVED** [review]

File: src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java (hunk 11 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/movietheatre/Main.java (hunk 17 above)
Comment (Sashk52) [reply src/main/java/com/dev/movietheatre/Main.java]: done

**Pull request merged**